### PR TITLE
Backend/309/dynamic roles permissions

### PR DESCRIPTION
**Pull request by carltonsmith**
### This pull request changes...

Adds functionality for supporting roles and permissions, as well as Django Admin abilities to manage users, groups and permissions. The user `auth_check` now returns the user's roles and associated permissions.

## TO TEST

Our recommendation is for the Raft tech lead @carltonsmith to demonstrate the capabilities as this backend centered pull request can be rather technical to test. Here are the changes that can be demonstrated as a result of this pull request.
- [ ] After logging in as a privileged user, using a Rest Client browser extension, polling the previously existing `auth_check` endpoint now returns the logged in user's assigned roles and permissions.
- [ ] A new endpoint exists to list the existing roles in the system at `/v1/roles/` (This satisfies the AC "Preliminary (MVP) roles are populated in the database")
**Security Control Requirement Changes**
_These changes were made to address security control requirements_
- [ ] A user with built-in Django roles `Staff` and `SuperUser` is able to log in to the Django Admin console
- [ ] In the Django Admin console all users' data can be updated including first_name, last_name, email and assigned roles (called Groups in the Django Admin)
- [ ] Groups can be managed by updating permissions for each group
- [ ] All changes made in the Django Admin Console are logged and viewable within the console

### This pull request is ready to merge when...

- [ ] Meets acceptance criteria for the issue 
- [ ] Code is reviewed by someone other than the original author
- [ ] Code is tested 
- [ ] Experience is approved by UX  
- [ ] Meets [a11y checklist](https://github.com/HHS/TANF-app/blob/main/docs/a11y/how-18f-will-test-a11y.md)
- [ ] Meets [Raft's Manual QASP Checklist](https://teams.microsoft.com/l/file/B5285301-13CA-46FD-A723-AFFD578C224A?tenantId=d58addea-5053-4a80-8499-ba4d944910df&fileType=docx&objectUrl=https%3A%2F%2Fhhsgov.sharepoint.com%2Fsites%2FTANFDataPortalOFA%2FShared%20Documents%2FGeneral%2FManual%20Testing%20Practices_Issue173.docx&baseUrl=https%3A%2F%2Fhhsgov.sharepoint.com%2Fsites%2FTANFDataPortalOFA&serviceName=teams&threadId=19:f769bbcb029f4f02b55ae7fad90e310d@thread.skype&groupId=41f194a6-c1d3-4680-933e-c8ee7d17e287):lock: (only applicable for PR to main HHS)
- [ ] Documentation is updated 
  - OpenAPI  
  - Readme
  - Entity Relationship Diagram (ERD)

### CODE REVIEW

**Comment by alexsoble** [comment]
Kicking the tires on the HHS CI.

File: tdrs-backend/docs/api/roles.md
Comment (alexsoble) [inline tdrs-backend/docs/api/roles.md:3]: This doc looks out of date to me -- my understanding is that admin users are no longer able to create/update roles via the API since that will now be done via the Django Admin dashboard.

File: tdrs-backend/docs/jwt_key_management.md
Comment (alexsoble) [inline tdrs-backend/docs/jwt_key_management.md:8]: This documentation looks incomplete. Since it's not strictly related to roles and permissions, we could break out documentation of JWT keys into a separate PR.

File: tdrs-backend/docs/security-tracking/sprint4_frontend_09292020.md
Comment (alexsoble) [inline tdrs-backend/docs/security-tracking/sprint4_frontend_09292020.md:1]: While I don't have any objections to merging this in, just noting that Sprint 4 was a while back, and I don't believe we're writing up reports for every Synk update anymore (see #113). So these update docs could be safely removed from my point of view.

File: tdrs-backend/docs/user_role_management.md
Comment (alexsoble) [inline tdrs-backend/docs/user_role_management.md:4]: Is this documentation up to date with the latest reflected in the PR?

File: tdrs-backend/tdpservice/core/models.py
Comment (alexsoble) [inline tdrs-backend/tdpservice/core/models.py:7]: @carltonsmith We discussed this last week — it's not clear to me if global permissions are being used at all in the application right now. 

If global permissions are being used, could you point me to where in the app they're being used?

If not, I think we should back this out and remove the files in this PR that introduced global permissions.

File: tdrs-backend/tdpservice/users/models.py
Comment (alexsoble) [inline tdrs-backend/tdpservice/users/models.py:24]: @carltonsmith Let's clean up TODO comments before QASP review / merge.

File: tdrs-backend/tdpservice/users/permissions.py
Comment (alexsoble) [inline tdrs-backend/tdpservice/users/permissions.py:11]: @carltonsmith What is the left-hand `request.user` side of the or statement doing here? I read it as saying that any user will have a `IsUserOrAdmin` value of `true`, since `request.user` can be truthy even if the user is not authenticated or not an admin.

File: tdrs-backend/tdpservice/users/test/test_api.py
Comment (alexsoble) [inline tdrs-backend/tdpservice/users/test/test_api.py:354]: @carltonsmith We talked last week about splitting this file up into smaller files since it's hard to track changes over large files.

File: tdrs-backend/tdpservice/users/views.py
Comment (alexsoble) [inline tdrs-backend/tdpservice/users/views.py:36]: I read this as saying that anyone can create a user object via the API without authentication, which doesn't seem like what we'd want. 

Since we have tests showing that unauthorized users can't create roles or permissions via the API, could we add a similar test for users to lock this down?

**Review by alexsoble — CHANGES_REQUESTED** [review]
@carltonsmith Took a first pass at a QASP review and left notes on a few areas where I have outstanding questions or comments, mostly around documentation and some of the permissions functionality.

**Comment by alexsoble** [comment]
@carltonsmith Less important than the comments above about documentation/functionality, but wanted to also leave notes on two binary file additions:

- `.DS_Store` -- looks from the `.gitignore` like this should be removed
- `tdp-erd.png` -- this is stored in a different folder in the HHS repo so I believe this PR would introduce a duplicate

File: tdrs-backend/docs/api/roles.md
Comment (carltonsmith) [reply tdrs-backend/docs/api/roles.md:3]: That's correct. I missed that the `POST` endpoint is still mentioned here.

File: tdrs-backend/docs/jwt_key_management.md
Comment (carltonsmith) [reply tdrs-backend/docs/jwt_key_management.md:8]: Makes sense

File: tdrs-backend/docs/security-tracking/sprint4_frontend_09292020.md
Comment (carltonsmith) [reply tdrs-backend/docs/security-tracking/sprint4_frontend_09292020.md:1]: I can remove it - this was something created much earlier

File: tdrs-backend/docs/user_role_management.md
Comment (carltonsmith) [reply tdrs-backend/docs/user_role_management.md:4]: That makes sense - updating with the Django Admin info

File: tdrs-backend/tdpservice/core/models.py
Comment (carltonsmith) [reply tdrs-backend/tdpservice/core/models.py:7]: @alexsoble While we did talk about creating endpoints for permissions, I don't recall discussing this directly. This is not to support permission endpoints, but to support any permissions that are not directly related to a specific model. Django, by default, has permissions for things such as "Can edit User" and "Can create STT", but for any permissions that are not directly related to a model, such as being able to view partial pages, not seeing particular links and such, global permissions are required. This is the base implementation to allow for that down the road.

Perhaps it is just a question of the approach one takes to programming, but even in an agile environment, I prefer to include the basic parts of something I know will be built upon and expanded in the future, because I know the basic architecture is there. I have run into issues in the past where leaving out the basics causes problems down the road. In my view, this is one of those circumstances and I would really prefer to not leave it out at this point.

File: tdrs-backend/tdpservice/users/models.py
Comment (carltonsmith) [reply tdrs-backend/tdpservice/users/models.py:24]: Yes, that makes sense

File: tdrs-backend/tdpservice/users/permissions.py
Comment (carltonsmith) [reply tdrs-backend/tdpservice/users/permissions.py:11]: @alexsoble I can see how it would read that way, but that's actually not what's happening here. This is a specific method of the Django Rest Framework to check on permissions for the object defined in the app. Since this `permissions.py` exist in the `users` app, the `obj` passed to this function is the user related to the object. So, say I am going to edit a user. When I do, this method will be passed an `obj` argument that represents the user I am trying to access. The check of `if obj == request.user` is asking if the object I am requesting access to and the user making the request is the same. If it is, that means that the user is attempting to edit their own profile.

If we were to include a `permissions.py` in another object and check for something similar, we could attach an `owner` to the `obj` and then check for `if obj.owner == request.user` the same way, but since the `obj` itself is the user here, this is the way it works.

File: tdrs-backend/tdpservice/users/test/test_api.py
Comment (carltonsmith) [reply tdrs-backend/tdpservice/users/test/test_api.py:354]: Yes, we did, and I definitely think it's a great idea. I just think it might be a bit heavier of a lift to attach to this PR. Would you be amenable to me creating a separate issue to break these up?

File: tdrs-backend/tdpservice/users/views.py
Comment (carltonsmith) [reply tdrs-backend/tdpservice/users/views.py:36]: @alexsoble Actually, this is how we are able to create an account for a user who doesn't yet exist in the system. Without this, the first time a user logged in to the system, it would block the system from creating a profile. Since one of our requirements was to automatically create a profile when the user logs in the first time, this is necessary to allow it.

File: tdrs-backend/docs/api/roles.md
Comment (carltonsmith) [reply tdrs-backend/docs/api/roles.md]: @alexsoble This is now updated accordingly

File: tdrs-backend/docs/jwt_key_management.md
Comment (carltonsmith) [reply tdrs-backend/docs/jwt_key_management.md]: @alexsoble This has been removed from the PR

File: tdrs-backend/docs/security-tracking/sprint4_frontend_09292020.md
Comment (carltonsmith) [reply tdrs-backend/docs/security-tracking/sprint4_frontend_09292020.md]: @alexsoble This has been removed from the PR

File: tdrs-backend/docs/user_role_management.md
Comment (carltonsmith) [reply tdrs-backend/docs/user_role_management.md]: @alexsoble This has been updated

File: tdrs-backend/tdpservice/users/models.py
Comment (carltonsmith) [reply tdrs-backend/tdpservice/users/models.py]: @alexsoble This has been updated

**Comment by carltonsmith** [comment]
> @carltonsmith Less important than the comments above about documentation/functionality, but wanted to also leave notes on two binary file additions:
> 
> * `.DS_Store` -- looks from the `.gitignore` like this should be removed
> * `tdp-erd.png` -- this is stored in a different folder in the HHS repo so I believe this PR would introduce a duplicate

No problem. .DS_Store has been removed and so was the dup ERD

File: tdrs-backend/docs/user_role_management.md
Comment (lfrohlich) [inline tdrs-backend/docs/user_role_management.md:1]: typo (management)

File: tdrs-backend/docs/user_role_management.md
Comment (lfrohlich) [inline tdrs-backend/docs/user_role_management.md:80]: Please correct to OFA MVP
```suggestion
For the OTA MVP, we will need to assign the Django built-in roles of `superuser` and `staff` to the deployed applicataion.
```

File: tdrs-backend/docs/user_role_management.md
Comment (lfrohlich) [inline tdrs-backend/docs/user_role_management.md:7]: Please correct to "OFA MVP"
```suggestion
For the OTA MVP we will be assigning and updating application user roles through the
```

File: tdrs-backend/docs/jwt_key_management.md
Comment (alexsoble) [reply tdrs-backend/docs/jwt_key_management.md]: Thanks! Though I'm still seeing "to create and update  roles" when I look at the changed files, should we remove that as well?

I'm also seeing "Accepts GET and requests".

File: tdrs-backend/docs/security-tracking/sprint4_frontend_09292020.md
Comment (alexsoble) [reply tdrs-backend/docs/security-tracking/sprint4_frontend_09292020.md]: Thanks! I'm also seeing a `sprint5_backend_10092020.md` file in the same folder that can be removed as well.

File: tdrs-backend/docs/user_role_management.md
Comment (alexsoble) [reply tdrs-backend/docs/user_role_management.md]: 👍 👍

File: tdrs-backend/tdpservice/core/models.py
Comment (alexsoble) [reply tdrs-backend/tdpservice/core/models.py:7]: Agreed that this is probably a difference in agile philosophy! The QASP doesn't speak to this directly one way or the other, since you feel that this will help future devs I'm happy to keep it in. 😄

Comment (alexsoble) [reply tdrs-backend/tdpservice/core/models.py:7]: My only thought is, might be worth adding a code comment explaining that this model isn't used yet for future developers trying to understand the permission structure

File: tdrs-backend/docs/jwt_key_management.md
Comment (alexsoble) [reply tdrs-backend/docs/jwt_key_management.md]: Oh sorry, left the above comment in the wrong thread!

Comment (alexsoble) [reply tdrs-backend/docs/jwt_key_management.md]: My only thought here is that we should update the tdrs-backend/README documentation that references the JWT key documentation if we remove the JWT key documentation

File: tdrs-backend/docs/api/roles.md
Comment (alexsoble) [reply tdrs-backend/docs/api/roles.md]: Thanks! Though I'm still seeing "to create and update  roles" when I look at the changed files, should we remove that as well?

I'm also seeing "Accepts GET and requests".

File: tdrs-backend/tdpservice/users/permissions.py
Comment (alexsoble) [reply tdrs-backend/tdpservice/users/permissions.py:11]: I see! Thanks for the explanation. So is this in reference to any ACs or use cases that we want to build out, or is more about preparing for future functionality in general?

File: tdrs-backend/tdpservice/users/views.py
Comment (alexsoble) [reply tdrs-backend/tdpservice/users/views.py:36]: Hmmmm, when I put `"create": [AllowAny]` on its own line locally and then comment it out and run the test suite locally, I don't see any additional test failures — do you see the same thing?

File: tdrs-backend/tdpservice/users/test/test_api.py
Comment (alexsoble) [reply tdrs-backend/tdpservice/users/test/test_api.py:354]: Hmmmm... 

On the one hand, I'm not sure if I'd want this improvement by itself to hold up the entire PR. 

On the other hand, moving new tests to a separate file instead of adding them to the bottom of the existing file should be a very straightforward change.

We did say we want to emphasize quality when it comes to the "iron triangle." 

And this speaks to an element in the [Heuristics](https://github.com/HHS/TANF-app/blob/cb15595d3ad6b7a984f083c8c81edde036315b59/docs/Technical-Documentation/Heuristics.md) doc:

Indicator | Good sign
-- | --
Code has logical organization. | When files become large they are broken out into a folder, and folder structure could be reasonably understood by a developer less-familiar with the project. Non dev: The code is well organized by file and folder.

File: tdrs-backend/tdpservice/users/models.py
Comment (alexsoble) [reply tdrs-backend/tdpservice/users/models.py]: 👍

File: tdrs-backend/docs/api/roles.md
Comment (carltonsmith) [reply tdrs-backend/docs/api/roles.md]: Yep, it's all set now

File: tdrs-backend/docs/security-tracking/sprint4_frontend_09292020.md
Comment (carltonsmith) [reply tdrs-backend/docs/security-tracking/sprint4_frontend_09292020.md]: Ok, that's been removed also

File: tdrs-backend/tdpservice/core/models.py
Comment (carltonsmith) [reply tdrs-backend/tdpservice/core/models.py:7]: This is all set

File: tdrs-backend/tdpservice/users/test/test_api.py
Comment (carltonsmith) [reply tdrs-backend/tdpservice/users/test/test_api.py:354]: Yeah, that makes sense. My initial thought was to do something that would take more work, but I believe all of these tests are related to the user app, so it shouldn't be as much effort as I was thinking.

File: tdrs-backend/tdpservice/users/permissions.py
Comment (carltonsmith) [reply tdrs-backend/tdpservice/users/permissions.py:11]: This is more about shoring up proper permissions for existing functionality. When we implemented creating users we hadn't implemented permissions yet. This makes sure what we already have is secured properly with the new functionality this PR adds.

File: tdrs-backend/docs/user_role_management.md
Comment (carltonsmith) [reply tdrs-backend/docs/user_role_management.md]: Thanks, this has been updated

File: tdrs-backend/docs/user_role_management.md
Comment (carltonsmith) [reply tdrs-backend/docs/user_role_management.md]: Thanks, this has been updated

File: tdrs-backend/docs/user_role_management.md
Comment (carltonsmith) [reply tdrs-backend/docs/user_role_management.md]: Thanks, this has been updated

File: tdrs-backend/tdpservice/users/test/test_api.py
Comment (carltonsmith) [reply tdrs-backend/tdpservice/users/test/test_api.py]: This is all set 😄

File: tdrs-backend/tdpservice/users/views.py
Comment (carltonsmith) [reply tdrs-backend/tdpservice/users/views.py]: @alexsoble Good catch on this one! This is now updated to what it should be, and will error on tests if you take out the "AllowAny" - The way it was set up it turned out a DRF base permission class was returning that it had permission instead of it checking our class. It's all working as expected now.

File: tdrs-backend/docs/api/roles.md
Comment (alexsoble) [reply tdrs-backend/docs/api/roles.md]: Looks good!

File: tdrs-backend/docs/security-tracking/sprint4_frontend_09292020.md
Comment (alexsoble) [reply tdrs-backend/docs/security-tracking/sprint4_frontend_09292020.md]: 👍  Thank you Carl!

File: tdrs-backend/tdpservice/core/models.py
Comment (alexsoble) [reply tdrs-backend/tdpservice/core/models.py:7]: Great!

File: tdrs-backend/tdpservice/users/test/test_api.py
Comment (alexsoble) [reply tdrs-backend/tdpservice/users/test/test_api.py]: Love it, much easier to read!

File: tdrs-backend/docs/jwt_key_management.md
Comment (carltonsmith) [reply tdrs-backend/docs/jwt_key_management.md]: Makes sense - this is done

Comment (alexsoble) [reply tdrs-backend/docs/jwt_key_management.md]: Hmmm, I'm still seeing it when I look at /files

File: tdrs-backend/tdpservice/users/views.py
Comment (alexsoble) [reply tdrs-backend/tdpservice/users/views.py]: Gotcha! When I look at `login.py`, here's the method I see that's creating new users:

```
    def handle_user(self, request, id_token, decoded_payload):
        """Handle the incoming user."""
        # get user from database if they exist. if not, create a new one
        if "token" not in request.session:
            request.session["token"] = id_token

        user = CustomAuthentication.authenticate(
            self, username=decoded_payload["email"]
        )
        if user is not None:
            self.login_user(request, user, "User Found")
        else:
            User = get_user_model()
            user = User.objects.create_user(decoded_payload["email"])
            user.set_unusable_password()
            user.save()
            self.login_user(request, user, "User Created")
```

It doesn't look like this method is hitting the API endpoint. 

So if we remove the API endpoint, would the create-new-user functionality break or stay the same? If I'm not looking at the right file or method please point me to the right place in the code.

Comment (alexsoble) [reply tdrs-backend/tdpservice/users/views.py]: Zooming out one level, I have a question about the longer mapping of actions to permissions classes in 1c5b4f8f6ad7851732ffa84fa6fc57b3dd7598c7. 

Do we want to be hard-coding what permissions belong to what groups? 

My understanding is that would be set or editable via the Django Admin interface, e.g.: 

screenshot of Django Admin dashboard

Comment (alexsoble) [reply tdrs-backend/tdpservice/users/views.py]: Hmmmm, this is starting to clear up for me a bit. 

There are two different sources of permissions being used in this app: permissions built in to Django, and permissions built into Django Rest Framework.

My understanding is that the Django permissions are stored in the database and can be edited via the Django Admin Interface, while the Django Rest Framework permissions aren't stored in the DB and aren't editable right now, is that correct?

File: tdrs-backend/docs/user_role_management.md
Comment (alexsoble) [inline tdrs-backend/docs/user_role_management.md:21]: @carltonsmith A couple of questions about this:

+ This will be different from the user's Login.gov password, is that correct? How does this interact with the `user.set_unusable_password()` call made from `login.py`?
+ This isn't protected by 2FA in the proposed PR, is that correct?

File: tdrs-backend/docs/jwt_key_management.md
Comment (carltonsmith) [reply tdrs-backend/docs/jwt_key_management.md]: Not sure what happened there, but it's updated now

File: tdrs-backend/tdpservice/users/views.py
Comment (carltonsmith) [reply tdrs-backend/tdpservice/users/views.py]: That is one way of looking at it - Django Model Permissions are loaded to the database automatically and manage CRUD permissions for all models, while DRF has some permissions you can use in the code such as `AllowAny` and `IsAuthenticated`.

I think that our long term strategy will be to implement default permissions leveraging the Django Model Permissions as it will give us a more granular approach to permissions, but implementing that is outside the scope of this issue.

Comment (carltonsmith) [reply tdrs-backend/tdpservice/users/views.py]: > So if we remove the create user API endpoint, would the create-new-user-on-login functionality break or stay the same? If I'm not looking at the right file or method please point me to the right place in the code.

I believe the "create-new-user-on-login" functionality would most likely remain the same, but I'd want to test it to be sure, if it's a priority.

Comment (carltonsmith) [reply tdrs-backend/tdpservice/users/views.py]: > Do we want to be hard-coding what permissions belong to what groups?
> 
> My understanding is that would be set or editable via the Django Admin interface, e.g.:

No, we don't want to hard code what permissions belong to what groups. Using the Django Admin is definitely a better approach for assigning permissions to groups.

File: tdrs-backend/docs/jwt_key_management.md
Comment (alexsoble) [reply tdrs-backend/docs/jwt_key_management.md]: 👍 👍

File: tdrs-backend/docs/user_role_management.md
Comment (carltonsmith) [reply tdrs-backend/docs/user_role_management.md:21]: - Yes, this will be different from the user's login.gov password. It doesn't interfere with the `user.set_unusable_password()` call. That call is necessary for Django to create the account, but doesn't break when the user's password is set for Django Admin.

- That's correct. I don't know if there is a way to turn on MFA for the admin interface but I can look into it.

File: tdrs-backend/tdpservice/users/views.py
Comment (alexsoble) [reply tdrs-backend/tdpservice/users/views.py]: Just noting that we documented removing the "create" endpoint in a separate issue here: 

https://app.zenhub.com/workspaces/tdrs-product-backlog-5f2c6cdc7c0bb1001bdc43a5/issues/raft-tech/tanf-app/518

File: tdrs-backend/docs/user_role_management.md
Comment (alexsoble) [reply tdrs-backend/docs/user_role_management.md:21]: Noting that we broke this out into a separate issue: https://app.zenhub.com/workspaces/tdrs-product-backlog-5f2c6cdc7c0bb1001bdc43a5/issues/raft-tech/tanf-app/517

**Comment by carltonsmith** [comment]
closing to create new PR with only needed updates

**Pull request closed without merging**
